File: client/src/observability/sentry.ts
/**
 * Sentry SDK initialization for the React client (E159 Part 1c).
 *
 * Single entry point: {@link initSentry}. Called once from `main.tsx`
 * BEFORE the React tree mounts so early render errors are captured.
 *
 * Behaviour:
 *
 * - Reads `import.meta.env.VITE_SENTRY_DSN` (baked at build time by Vite).
 * - Reads `import.meta.env.VITE_GIT_SHA` for the Sentry `release` tag so
 *   stack traces deep-link back to the exact commit.
 * - Production + missing DSN → `console.warn` (does NOT throw). Unlike the
 *   server, an SPA must not refuse to render just because observability
 *   is absent — the user still needs the app.
 * - Dev / test with empty DSN → silent no-op.
 * - Sample rates: 10% in production, 100% in dev.
 * - `sendDefaultPii: false` — GDPR-safe default; opt-in only.
 */

import * as Sentry from "@sentry/react";

/** Initialise Sentry for the browser client. Idempotent — safe to call once. */
export function initSentry(): void {
  const dsn = import.meta.env.VITE_SENTRY_DSN as string | undefined;
  const isProd = import.meta.env.PROD;
  const mode = import.meta.env.MODE as string;
  const release = (import.meta.env.VITE_GIT_SHA as string | undefined) || "unknown";

  if (!dsn) {
    if (isProd) {
      // Production without a DSN is almost always a deploy mistake. Warn loudly
      // but do not throw — the app still has to render for end users.
      // eslint-disable-next-line no-console
      console.warn(
        "[observability] VITE_SENTRY_DSN is empty in a production build. " +
          "Set it in Zeabur (or your platform) BEFORE the client build runs.",
      );
    }
    return; // dev/test or missing prod DSN: silent no-op
  }

  Sentry.init({
    dsn,
    environment: mode,
    release,
    integrations: [Sentry.browserTracingIntegration()],
    tracesSampleRate: isProd ? 0.1 : 1.0,
    sendDefaultPii: false,
  });
}

File: server/app/observability/sentry.py
"""Sentry SDK initialization (E159 Part 1).

Single entry point: :func:`init_sentry`. Called once at app startup.

Behaviour summary:

- Reads ``SENTRY_DSN_SERVER`` (preferred) — falls back to legacy ``SENTRY_DSN``
  for backwards compatibility with existing deployments.
- ``ENVIRONMENT=production`` + empty DSN -> raises ``RuntimeError`` (fail-fast).
  All other environments are silent no-ops when the DSN is empty.
- ``release`` tag is taken from ``GIT_SHA`` env (else ``"unknown"``) so Sentry
  can deep-link stack traces to commits.
- Sample rates: 10% in production, 100% elsewhere. Profiling: 10% in prod,
  off in dev (saves battery during local work).
- ``send_default_pii=False`` — GDPR-safe default; opt-in only.
"""

from __future__ import annotations

import os

import sentry_sdk
from sentry_sdk.integrations.fastapi import FastApiIntegration
from sentry_sdk.integrations.sqlalchemy import SqlalchemyIntegration


def _resolve_dsn() -> str:
    """Prefer ``SENTRY_DSN_SERVER`` (E159 contract); fall back to ``SENTRY_DSN``.

    The legacy ``SENTRY_DSN`` env was the original contract (set by
    `app.main` before E159). Keeping the fallback prevents existing
    Zeabur deployments from breaking the moment this module ships.
    """

    return os.getenv("SENTRY_DSN_SERVER") or os.getenv("SENTRY_DSN", "")


def init_sentry() -> None:
    """Initialise the Sentry SDK for the server process.

    Raises:
        RuntimeError: when ``ENVIRONMENT=production`` and no DSN is configured.
            This is a hard fail because shipping prod without error tracking
            is the silent failure mode E159 was created to eliminate.
    """

    dsn = _resolve_dsn()
    env = os.getenv("ENVIRONMENT", "development")

    if env == "production" and not dsn:
        raise RuntimeError(
            "SENTRY_DSN_SERVER is required when ENVIRONMENT=production. "
            "Set it in Zeabur (or your platform secrets) before deploy. "
            "Local/staging may leave it empty for a silent no-op."
        )

    if not dsn:
        return  # dev/test: Sentry off — silent no-op

    sentry_sdk.init(
        dsn=dsn,
        environment=env,
        release=os.getenv("GIT_SHA", "unknown"),
        traces_sample_rate=0.1 if env == "production" else 1.0,
        profiles_sample_rate=0.1 if env == "production" else 0.0,
        integrations=[FastApiIntegration(), SqlalchemyIntegration()],
        send_default_pii=False,
    )
